src/base/detail/Joined.hh

```diff
@@ -0,0 +1,98 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file Joined.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <iterator>
+#include <iosfwd>
+#include <sstream>
+#include <string>
+
+namespace celeritas
+{
+namespace detail
```

Comment:
Do we already have (and/or have the need for) a distinction between a ```detail``` sub-namespace (entities that the user might use but are secondary interface and should be used only by 'advanced' users) and a ```internal``` sub-namespace (entities that no user should use, either because they are too low level to use from the outside or because they might change every release or just because we don't want to support their usage outside the library).  [related :) .. is the current usage of ```detail``` the first or 2nd case?]

Reply:
My intention for the `detail` namespace is that they're implementation details used only usable by classes in the immediate parent directory. (As a detail they can be shared between multiple classes if needed, and they can also be tested independently. And it can be used to split NVCC-built code from regular host compiler code.)

Reply:
fair enough (so it is closer that what I was calling "internal").